src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/MemorySegmentPool.java

```diff
@@ -0,0 +1,201 @@
+ * 1GB + 1, pool will allocate 2GB of memory.
+ */
+public class MemorySegmentPool {
+  private static final int[] DEFAULT_MAX_SIZES = new int[Long.SIZE];
+  private static final ResourceScope GLOBAL = ResourceScope.globalScope();
+
+  @SuppressWarnings({"rawtypes", "unchecked"})
+  private final SpinLockQueue<MemoryPoolSegment> segmentsDequeue[] = new SpinLockQueue[Long.SIZE];
+
+  private final ResourceScope scope;
+
+  static {
+    int idx = 0;
+    int cores = Runtime.getRuntime().availableProcessors();
+
+    for (; idx <= 10; idx++) {
+      DEFAULT_MAX_SIZES[idx] = 256;
+    }
+
+    // Sizes up to 64kb
+    for (; idx <= 16; idx++) {
+      DEFAULT_MAX_SIZES[idx] = cores * 4;
+    }
+
+    // Sizes up to 1MB {
+    for (; idx <= 20; idx++) {
+      DEFAULT_MAX_SIZES[idx] = cores;
+    }
+
+    DEFAULT_MAX_SIZES[idx++] = Math.min(cores, 2); //2mb
+    DEFAULT_MAX_SIZES[idx++] = Math.min(cores, 1); //4mb
+    // Rest 0
+  }
+
+  public MemorySegmentPool(ResourceScope scope) {
+    this(DEFAULT_MAX_SIZES, scope);
+  }
+  public MemorySegmentPool(int maxSizes[], ResourceScope scope) {
+    this.scope = scope;
+
+    validateMaxSizes(maxSizes);
+    for (int i=0; i < segmentsDequeue.length; i++) {
+      segmentsDequeue[i] = new SpinLockQueue<>(maxSizes[i]);
+    }
+  }
+
+  public SegmentAllocator allocatorForScope(ResourceScope resourceScope) {
+    // Prevent scope managing this pool to go away, when dependant allocator is alive
+    final var handle = scope.acquire();
+    resourceScope.addOnClose(handle::close);
+    return (bytesSize, bytesAlignment) -> getSegmentForScope(resourceScope, bytesSize, bytesAlignment);
+  }
+
+  public MemoryPoolSegment getSegmentEntryByLayout(MemoryLayout layout) {
+    return getSegmentEntryBySize(layout.byteSize(), layout.byteAlignment());
+  }
+
+  /**
+   * Gets segment from pool or allocates new one. Internally segments are cached.
+   * The size of segment can be larger than requested.
+   *
+   * @param size the size of segment.
+   *
+   * @return segment of size at least `size`
+   */
+  @ForceInline
+  public MemoryPoolSegment getSegmentEntryBySize(long size, long alignment) {
+    if (!scope.isAlive()) {
+      throw new IllegalStateException("Associated resource scope is closed");
+    }
+
+    final var alignedSize = (size + alignment - 1) & -alignment;
+    final var bitBound = bitBound(alignedSize);
+    final var segmentDequeue = segmentsDequeue[bitBound];
+
+    var segment = segmentDequeue.pollEntry();
+    if (segment == null) {
+      final var bitBoundedSize = 1L << bitBound;
+      segment = allocateNewEntry(segmentDequeue, bitBoundedSize);
+    }
+
+    return segment;
+  }
+
+  private MemorySegment getSegmentForScope(ResourceScope resourceScope, long size, long alignment) {
+    final var segmentEntry = getSegmentEntryBySize(size, alignment);
+    resourceScope.addOnClose(() -> segmentEntry.close());
```

Comment:
There's a potential safety issue here - resourceScope doesn't promise that cleanup actions will be called in a certain order. But I think this should be ok, because the cleanup actions are invoked _after_ the user-provided scope is closed. Meaning that the segments associated with that scope will not be accessible anyway.

Reply:
Good catch, in case when scope is getting closed in fact segment can be abandoned, and will be permanently removed from pool.